src/gwt/src/org/rstudio/core/client/command/AppCommand.java

```diff
@@ -157,9 +157,6 @@ private void doExecute()
       assert enabled_ : "AppCommand executed when it was not enabled";
       if (!enabled_)
          return;
-      assert visible_ : "AppCommand executed when it was not visible";
```

Comment:
Without this check you can execute commands that have been removed. For example, on open-source server (and desktop) we remove the command for returning to the Workbench home page:

https://github.com/rstudio/rstudio/blob/2d2dcd9db6f3ee0e15903ae377cf89931fc1f2d0/src/gwt/src/org/rstudio/studio/client/application/Application.java#L1032

Prior to this change `.rs.api.executeCommand("loadServerHome")` in the console did nothing on open-source, but with this change you'll be navigated to a semi-broken page on RStudio Server, and on Desktop it navigates into a broken state.

I don't know if this would cause any actual end-user problems, though. This might be in the "don't poke yourself in the eye with a stick if you don't want a stick in your eye" scenario.

Reply:
There's probably a couple options though:

- Have `doExecute()` check the `removed_` flag, and disallow execution of removed commands?
- Have `remove()` also set `enabled_ = false`?

I also realized we have this bit of "smelly" code:

https://github.com/rstudio/rstudio/blob/2d2dcd9db6f3ee0e15903ae377cf89931fc1f2d0/src/gwt/src/org/rstudio/core/client/command/AppCommand.java#L205

so there's definitely been some amount of internal confusion / uncertainty around handling "enabled-but-not-visible" commands :-)

Reply:
If `.rs.api.executeCommand` is the only way to invoke these "removed" commands, it's probably fine as-is.

Reply:
I'm sure Joe remembers EXACTLY what he had in mind with this line of code 14 years ago.